File: apps/control-api/app/rulepack.py
"""威胁检测规则包加载：内置包 + 可选的外部签名更新包（P2）。

安全红线（fail-closed）：
- 外部规则包（SIQ_AS_THREAT_RULEPACK_PATH 指向的本地 JSON，即离线缓存）必须带
  控制面 Ed25519 签名：旁边同名 `<文件名>.sig` 文件，base64 编码，签名输入 =
  规则包 JSON 的规范序列化（sort_keys + 紧凑分隔符，与 app.signing 一致）；
- 验签失败 / 缺 .sig / JSON 畸形 / 字段缺漏 / 正则不可编译 / 版本低于内置包，
  一律拒绝并回退内置包，绝不加载未验签或降级的规则；
- 拒绝原因只记类别与路径，规则包内容不进日志；
- 内置包损坏是构建事故：直接 RuntimeError 拒绝启动，不静默。
"""

from __future__ import annotations

import base64
import json
import logging
import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from cryptography.exceptions import InvalidSignature

from app.signing import _canonical_bytes, get_signing_key

logger = logging.getLogger(__name__)

# 外部更新包路径（离线缓存 = 本地文件路径；回滚 = 指回旧版本文件或移除该变量）
RULEPACK_PATH_ENV = "SIQ_AS_THREAT_RULEPACK_PATH"

_BUILTIN_PATH = Path(__file__).resolve().parent / "data" / "threat_rules.v1.json"

_SEVERITIES = frozenset({"critical", "high", "medium", "low"})
_REQUIRED_FIELDS = ("rule_id", "severity", "confidence", "description", "impact", "remediation", "patterns")


@dataclass(frozen=True)
class Rule:
    rule_id: str
    severity: str  # critical|high|medium|low
    confidence: float  # 0-1
    description: str
    impact: str
    remediation: str
    patterns: tuple[re.Pattern[str], ...]


class RulepackError(ValueError):
    """规则包非法/验签失败。消息只含类别信息，不含规则内容（可安全进日志）。"""


def _parse_rulepack(data: Any) -> tuple[int, tuple[Rule, ...]]:
    """校验规则包结构并编译 patterns；任何非法都抛 RulepackError（整包拒绝）。"""
    if not isinstance(data, dict):
        raise RulepackError("rulepack root must be an object")
    version = data.get("version")
    if not isinstance(version, int) or isinstance(version, bool) or version < 1:
        raise RulepackError("rulepack 'version' must be a positive integer")
    raw_rules = data.get("rules")
    if not isinstance(raw_rules, list) or not raw_rules:
        raise RulepackError("rulepack 'rules' must be a non-empty array")
    rules: list[Rule] = []
    seen_ids: set[str] = set()
    for idx, raw in enumerate(raw_rules):
        if not isinstance(raw, dict):
            raise RulepackError(f"rule #{idx} must be an object")
        missing = [f for f in _REQUIRED_FIELDS if f not in raw]
        if missing:
            raise RulepackError(f"rule #{idx} missing required field(s): {', '.join(missing)}")
        rule_id = raw["rule_id"]
        if not isinstance(rule_id, str) or not rule_id or rule_id in seen_ids:
            raise RulepackError(f"rule #{idx} has invalid or duplicate rule_id")
        seen_ids.add(rule_id)
        if raw["severity"] not in _SEVERITIES:
            raise RulepackError(f"rule #{idx} has invalid severity")
        confidence = raw["confidence"]
        if not isinstance(confidence, (int, float)) or isinstance(confidence, bool) or not 0 < confidence <= 1:
            raise RulepackError(f"rule #{idx} has invalid confidence")
        if not all(isinstance(raw[f], str) for f in ("description", "impact", "remediation")):
            raise RulepackError(f"rule #{idx} has non-string text field")
        raw_patterns = raw["patterns"]
        if (
            not isinstance(raw_patterns, list)
            or not raw_patterns
            or not all(isinstance(p, str) for p in raw_patterns)
        ):
            raise RulepackError(f"rule #{idx} 'patterns' must be a non-empty array of strings")
        try:
            patterns = tuple(re.compile(p) for p in raw_patterns)
        except re.error as exc:
            raise RulepackError(f"rule #{idx} contains an uncompilable pattern") from exc
        rules.append(
            Rule(
                rule_id=rule_id,
                severity=raw["severity"],
                confidence=float(confidence),
                description=raw["description"],
                impact=raw["impact"],
                remediation=raw["remediation"],
                patterns=patterns,
            )
        )
    return version, tuple(rules)


def _load_builtin() -> tuple[int, tuple[Rule, ...]]:
    """加载内置包；损坏即构建事故，RuntimeError 拒绝启动（不静默回退）。"""
    try:
        data = json.loads(_BUILTIN_PATH.read_text(encoding="utf-8"))
        return _parse_rulepack(data)
    except (OSError, json.JSONDecodeError, RulepackError) as exc:
        raise RuntimeError(f"内置威胁规则包损坏或缺失（{_BUILTIN_PATH.name}），属构建事故，拒绝启动: {exc}") from exc


def _verify_signature(path: Path, data: dict) -> None:
    """用控制面签名公钥验签 `<文件名>.sig`；任何失败抛 RulepackError。"""
    sig_path = path.with_name(path.name + ".sig")
    try:
        sig_b64 = sig_path.read_text(encoding="ascii").strip()
    except OSError as exc:
        raise RulepackError("missing signature file (.sig)") from exc
    try:
        signature = base64.b64decode(sig_b64, validate=True)
    except ValueError as exc:
        raise RulepackError("signature file is not valid base64") from exc
    try:
        get_signing_key().public_key().verify(signature, _canonical_bytes(data))
    except InvalidSignature as exc:
        raise RulepackError("signature verification failed") from exc


def load_rulepack() -> tuple[int, tuple[Rule, ...]]:
    """加载生效规则包，返回 (version, rules)。可重复调用（热更新/测试）。

    默认内置包；SIQ_AS_THREAT_RULEPACK_PATH 指向的外部包经验签通过且
    version >= 内置版本才采用，否则回退内置包并 logging.warning（fail-closed）。
    """
    builtin = _load_builtin()
    raw_path = os.getenv(RULEPACK_PATH_ENV)
    if not raw_path:
        return builtin
    path = Path(raw_path)
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        external = _parse_rulepack(data)
        _verify_signature(path, data)
    except (OSError, json.JSONDecodeError, RulepackError) as exc:
        logger.warning("外部威胁规则包被拒绝，回退内置包: path=%s reason=%s", path, exc)
        return builtin
    except RuntimeError as exc:  # 签名密钥不可用：fail-closed，不加载未验签规则
        logger.warning("外部威胁规则包验签密钥不可用，回退内置包: path=%s reason=%s", path, exc)
        return builtin
    if external[0] < builtin[0]:
        logger.warning(
            "外部威胁规则包版本低于内置包被拒绝（防回滚降级），回退内置包: path=%s", path
        )
        return builtin
    return external

File: apps/agentshield/internal/rulepack/rulepack.go
// Package rulepack loads the signed threat-detection rule pack shared with the
// Python control plane (apps/control-api/app/rulepack.py).
//
// Fail-closed rules mirror the Python side:
//   - the embedded built-in pack is the baseline; a corrupt built-in is a build
//     defect and Load returns an error instead of degrading;
//   - an external pack is used only if it parses, its <file>.sig (base64
//     Ed25519 over the canonical JSON bytes) verifies against the configured
//     public key, and its version is >= the built-in version;
//   - any rejection falls back to the built-in pack and reports only a reason
//     category, never rule contents.
//
// Regex dialect: Go RE2. The shared JSON is kept RE2-compatible (ADR-011 D2);
// a pattern that fails to compile rejects the whole pack.
package rulepack

import (
	"crypto/ed25519"
	_ "embed"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"regexp"
	"strings"

	"siq-agent-security/apps/agentshield/internal/canon"
)

//go:embed data/threat_rules.v1.json
var builtinJSON []byte

// PathEnv mirrors SIQ_AS_THREAT_RULEPACK_PATH on the Python side.
const PathEnv = "SIQ_AS_THREAT_RULEPACK_PATH"

var severities = map[string]bool{"critical": true, "high": true, "medium": true, "low": true}

// Rule is one detection rule with compiled patterns.
type Rule struct {
	ID          string
	Severity    string
	Confidence  float64
	Description string
	Impact      string
	Remediation string
	Patterns    []*regexp.Regexp
}

// Redaction is one excerpt-redaction substitution.
type Redaction struct {
	Pattern     *regexp.Regexp
	Replacement string
}

// Pack is a validated rule pack.
type Pack struct {
	Version    int
	Rules      []Rule
	Redactions []Redaction
	// Source is "builtin" or the external path that was accepted.
	Source string
}

// Error is a pack rejection. Its message is category-only and safe to log.
type Error struct{ msg string }

func (e *Error) Error() string { return e.msg }

func rejectf(format string, a ...any) error { return &Error{msg: fmt.Sprintf(format, a...)} }

// defaultRedactions mirrors _DEFAULT_REDACTION_RULES in rulepack.py and is
// used when an external pack omits redaction_patterns.
var defaultRedactions = []Redaction{
	{regexp.MustCompile(`sk-[a-zA-Z0-9_\-]{8,}`), "sk-***"},
	{regexp.MustCompile(`gh[pousr]_[a-zA-Z0-9]{20,}`), "gh_***"},
	{regexp.MustCompile(`(?i)bearer\s+[a-zA-Z0-9_\-\.]{10,}`), "Bearer ***"},
	{regexp.MustCompile(`(?i)password\s*=\s*['"][^'"]+['"]`), "password=***"},
	{regexp.MustCompile(`(?i)token\s*=\s*['"][^'"]+['"]`), "token=***"},
	{regexp.MustCompile(`(?i)secret\s*=\s*['"][^'"]+['"]`), "secret=***"},
}

// Parse validates and compiles a rule pack document.
func Parse(raw []byte) (*Pack, error) {
	var doc struct {
		Version   json.RawMessage   `json:"version"`
		Rules     []json.RawMessage `json:"rules"`
		Redaction json.RawMessage   `json:"redaction_patterns"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, rejectf("rulepack root must be an object")
	}
	var version int
	if err := json.Unmarshal(doc.Version, &version); err != nil || version < 1 || strings.ContainsAny(string(doc.Version), ".eE") {
		return nil, rejectf("rulepack 'version' must be a positive integer")
	}
	if len(doc.Rules) == 0 {
		return nil, rejectf("rulepack 'rules' must be a non-empty array")
	}
	seen := map[string]bool{}
	pack := &Pack{Version: version}
	for idx, rr := range doc.Rules {
		var r struct {
			ID          *string   `json:"rule_id"`
			Severity    *string   `json:"severity"`
			Confidence  *float64  `json:"confidence"`
			Description *string   `json:"description"`
			Impact      *string   `json:"impact"`
			Remediation *string   `json:"remediation"`
			Patterns    *[]string `json:"patterns"`
		}
		if err := json.Unmarshal(rr, &r); err != nil {
			return nil, rejectf("rule #%d must be an object with typed fields", idx)
		}
		if r.ID == nil || r.Severity == nil || r.Confidence == nil || r.Description == nil || r.Impact == nil || r.Remediation == nil || r.Patterns == nil {
			return nil, rejectf("rule #%d missing required field(s)", idx)
		}
		if *r.ID == "" || seen[*r.ID] {
			return nil, rejectf("rule #%d has invalid or duplicate rule_id", idx)
		}
		seen[*r.ID] = true
		if !severities[*r.Severity] {
			return nil, rejectf("rule #%d has invalid severity", idx)
		}
		if !(*r.Confidence > 0 && *r.Confidence <= 1) {
			return nil, rejectf("rule #%d has invalid confidence", idx)
		}
		if len(*r.Patterns) == 0 {
			return nil, rejectf("rule #%d 'patterns' must be a non-empty array of strings", idx)
		}
		rule := Rule{
			ID: *r.ID, Severity: *r.Severity, Confidence: *r.Confidence,
			Description: *r.Description, Impact: *r.Impact, Remediation: *r.Remediation,
		}
		for _, p := range *r.Patterns {
			re, err := regexp.Compile(p)
			if err != nil {
				return nil, rejectf("rule #%d contains an uncompilable pattern", idx)
			}
			rule.Patterns = append(rule.Patterns, re)
		}
		pack.Rules = append(pack.Rules, rule)
	}
	red, err := parseRedactions(doc.Redaction)
	if err != nil {
		return nil, err
	}
	pack.Redactions = red
	return pack, nil
}

func parseRedactions(raw json.RawMessage) ([]Redaction, error) {
	if len(raw) == 0 || string(raw) == "null" {
		return defaultRedactions, nil
	}
	var items []struct {
		Pattern     *string `json:"pattern"`
		Replacement *string `json:"replacement"`
	}
	if err := json.Unmarshal(raw, &items); err != nil || len(items) == 0 {
		return nil, rejectf("rulepack 'redaction_patterns' must be a non-empty array when present")
	}
	out := make([]Redaction, 0, len(items))
	for idx, it := range items {
		if it.Pattern == nil || *it.Pattern == "" {
			return nil, rejectf("redaction_patterns #%d missing non-empty string 'pattern'", idx)
		}
		if it.Replacement == nil {
			return nil, rejectf("redaction_patterns #%d missing string 'replacement'", idx)
		}
		re, err := regexp.Compile(*it.Pattern)
		if err != nil {
			return nil, rejectf("redaction_patterns #%d contains an uncompilable pattern", idx)
		}
		out = append(out, Redaction{re, *it.Replacement})
	}
	return out, nil
}

// Builtin returns the embedded pack. A parse failure is a build defect.
func Builtin() (*Pack, error) {
	p, err := Parse(builtinJSON)
	if err != nil {
		return nil, fmt.Errorf("built-in threat rulepack is corrupt (build defect): %w", err)
	}
	p.Source = "builtin"
	return p, nil
}

// BuiltinBytes exposes the embedded document for parity tests.
func BuiltinBytes() []byte { return builtinJSON }

// VerifySignature checks <path>.sig (base64 Ed25519) over the canonical bytes
// of the document at path.
func VerifySignature(pub ed25519.PublicKey, path string, raw []byte) error {
	sigB64, err := os.ReadFile(path + ".sig")
	if err != nil {
		return rejectf("missing signature file (.sig)")
	}
	sig, err := base64.StdEncoding.Strict().DecodeString(strings.TrimSpace(string(sigB64)))
	if err != nil {
		return rejectf("signature file is not valid base64")
	}
	v, err := canon.Decode(raw)
	if err != nil {
		return rejectf("rulepack is not valid JSON")
	}
	msg, err := canon.Marshal(v)
	if err != nil {
		return rejectf("rulepack cannot be canonicalised")
	}
	if len(pub) != ed25519.PublicKeySize || !ed25519.Verify(pub, msg, sig) {
		return rejectf("signature verification failed")
	}
	return nil
}

// Load returns the effective pack: the external pack named by PathEnv when it
// parses, verifies against pub and is not a downgrade; otherwise the built-in.
// warn receives a category-only reason whenever the external pack is rejected.
func Load(pub ed25519.PublicKey, warn func(string)) (*Pack, error) {
	builtin, err := Builtin()
	if err != nil {
		return nil, err
	}
	path := os.Getenv(PathEnv)
	if path == "" {
		return builtin, nil
	}
	if warn == nil {
		warn = func(string) {}
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		warn("external rulepack unreadable; using builtin")
		return builtin, nil
	}
	ext, err := Parse(raw)
	if err != nil {
		warn("external rulepack rejected (" + err.Error() + "); using builtin")
		return builtin, nil
	}
	if len(pub) == 0 {
		warn("external rulepack rejected (no verification key); using builtin")
		return builtin, nil
	}
	if err := VerifySignature(pub, path, raw); err != nil {
		warn("external rulepack rejected (" + err.Error() + "); using builtin")
		return builtin, nil
	}
	if isDowngrade(ext, builtin) {
		warn("external rulepack rejected (version downgrade); using builtin")
		return builtin, nil
	}
	ext.Source = path
	return ext, nil
}

// isDowngrade mirrors the Python anti-rollback rule: an external pack older
// than the built-in is refused; an equal version is accepted.
func isDowngrade(ext, builtin *Pack) bool { return ext.Version < builtin.Version }

// ErrRejected reports whether err is a pack rejection (as opposed to a build defect).
func ErrRejected(err error) bool {
	var e *Error
	return errors.As(err, &e)
}
